File: ctci/linkedlist/linkedlist.rb
def benchmark
  b = Time.now
  yield
  a = Time.now
  a - b
end

class LinkedList
  attr_reader :head
  Node = Struct.new :value, :next
  def initialize *nodes
    @head = Node.new
    nodes.each do |value|
      push value
    end
  end
  def partition val
    runner = @head
    while runner and runner.next.value do
      if runner.next.value < val
        push runner.next.value
        delete runner.next
      else
        runner = runner.next
      end
    end
  end
  def get idx, curr=@head,count=0
    return -1 unless curr
    return curr if count == idx
    get idx, curr.next, count + 1  
  end
  def delete node
    destroy = node.next.dup
    node.next = destroy.next
    node.value = destroy.value
  end
  def push value
    new_node = Node.new(value)
    new_node.next = @head.dup
    @head = new_node
  end
  def remove_dups curr=@head
    runner = curr
    while runner and runner.next do
      if runner.next.value == curr.value
        runner.next = runner.next.next
      else
        runner = runner.next
      end
    end
    curr.next ? remove_dups(curr.next) : return
  end
  def kth_to_last(curr=@head, k)
    loop do
      runner = curr
      k.times do
        runner = runner.next
      end
      break curr unless runner
      curr = curr.next
    end
  end
  def to_s list=@head, msg=""
    return msg[0..-4] if list.value.nil?
    to_s list.next, msg<<"#{list.value} -> "
  end
end

File: ctci/linkedlist/linkedlist.c
#include <stdio.h>
#include <stdlib.h>

typedef struct list {
  int val;
  struct list * next;
} Node;

void printList(Node * curr);

void push(Node ** headRef, int newValue);

void removeDups(Node * curr);

Node * kth_to_last_el(Node * curr, int k);

void deleteNode(Node * curr);

void partition(Node ** head, int val);

void main(){
  Node * head = (Node *)malloc(sizeof(Node));
  head->val = 5;
  head->next = NULL;
  push(&head, 4);
  push(&head, 3);
  push(&head, 2);
  push(&head, 1);

  printList(head);

  partition(&head, 4);

  printList(head);
}

void partition(Node ** head, int val){
  Node * runner = *head;
  while(runner->next && runner->next->val){
    if(runner->next->val < val){
      push(head, runner->next->val);
      deleteNode(runner->next);
    } else {
      runner = runner->next; 
    }
  }
}

void deleteNode(Node * curr){
  Node * destroy = curr->next;
  curr->val = destroy->val;
  curr->next = destroy->next; 
  free(destroy);
}

Node * kth_to_last_el(Node * curr, int k){
  int i;
  Node * runner = (Node *)malloc(sizeof(Node));
  for(;;){
    runner = curr;
    for(i = 0; i < k; i++) runner = runner->next;
    if(!runner) return curr;
    curr = curr->next;
  }
}

void push(Node ** headRef, int newValue){
  Node * newNode = (Node *)malloc(sizeof(Node));
  newNode->val = newValue;
  newNode->next = *headRef;
  *headRef = newNode;
}

void printList(Node * curr){
  if(!curr){
    printf("\n");
    return;
  }
  printf("%d ->", curr->val);
  printList(curr->next);
}

void removeDups(Node * curr){
  Node * runner = curr;
  Node * dup = (Node *)malloc(sizeof(Node *));
  while(runner && runner->next){
    if(runner->next->val == curr->val){
      dup = runner->next;
      runner->next = runner->next->next; 
      free(dup);
    } else {
      runner = runner->next;
    }
  }
  if(!curr->next) return;
  removeDups(curr->next);
}
